librad/src/id/entity.rs

```diff
@@ -0,0 +1,614 @@
+#[derive(Clone, Debug)]
+pub enum Signatory {
+    User(RadicleUri),
+    OwnedKey,
+}
+
+/// A signature for an `Entity`
+#[derive(Clone, Debug)]
+pub struct EntitySignature {
+    /// Who is producing this signature
+    pub by: Signatory,
+    /// The signature data
+    pub sig: Signature,
+}
+
+/// An URN resolver that turns URNs into `Entity` instances
+/// (`T` is the entity type)
+#[async_trait]
+pub trait Resolver<T> {
+    /// Resolve the given URN and deserialize the target `Entity`
+    async fn resolve(&self, uri: &RadicleUri) -> Result<T, Error>;
+}
+
+/// The base entity definition.
+///
+/// Entities have the following properties:
+///
+/// - They can evolve over time, so they have a sequence of revisions.
+/// - Their identity is stable (it does not change over time), and it is the
+///   hash of their initial revision.
+/// - Each revision contains the hash of the previous revision, which is also
+///   hashed, so that the sequence of revisions is a Merkel tree (actually just
+///   a list).
+/// - They can be signed, either with a key they own, or using a key belonging
+///   to a different entity (the certifier); note that when applying multiple
+///   signatures, signatures are not themselves signed (what is signed is always
+///   only the entity itself).
+/// - Each revision specifies the set of owned keys and trusted certifiers.
+/// - Each revision must be signed by all its owned keys and trusted certifiers.
+/// - Each subsequent revision must be signed by a quorum of the previous keys
+///   and certifiers, to prove that the entity evolution is actually under the
+///   control of its current "owners" (the idea is taken from TUF).
+#[derive(Clone)]
+pub struct Entity<T> {
+    /// The entity name (useful for humans because the hash is unreadable)
+    name: String,
+    /// Entity revision, to be incremented at each entity update
+    revision: u64,
+    /// Entity hash, computed on everything except the signatures and
+    /// (obviously) the hash itself
+    hash: Multihash,
+    /// Hash of the previous revision, `None` for the initial revision
+    /// (in this case the entity hash is actually the entity ID)
+    parent_hash: Option<Multihash>,
+    /// Set of signatures
+    signatures: HashMap<PublicKey, EntitySignature>,
+    /// Set of owned keys
+    keys: HashSet<PublicKey>,
+    /// Set of certifiers (entities identified by their URN)
+    certifiers: HashSet<RadicleUri>,
+    /// Specific `Entity` data
+    info: T,
+}
+
+impl<T> Entity<T>
+where
+    T: Serialize + DeserializeOwned + Clone + Default,
+{
+    /// `name` getter
+    pub fn name(&self) -> &str {
+        &self.name
+    }
+
+    /// `revision` getter
+    pub fn revision(&self) -> u64 {
+        self.revision
+    }
+
+    /// Build and `Entity` from its data (the second step of deserialization)
+    /// It guarantees that the `hash` is correct
+    pub fn from_data(data: data::EntityData<T>) -> Result<Self, Error> {
```

Comment:
Wouldn't this be better served as an impl of [TryFrom](https://doc.rust-lang.org/std/convert/trait.TryFrom.html)?